src/util/ssl_utils.js

```diff
@@ -27,26 +32,72 @@ function verify_ssl_certificate(certificate) {
     tls.createSecureContext(certificate);
 }
 
-function read_ssl_certificate() {
-    return P.resolve()
-        .then(() => P.props({
-            key: fs.readFileAsync(SERVER_SSL_KEY_PATH, 'utf8'),
-            cert: fs.readFileAsync(SERVER_SSL_CERT_PATH, 'utf8'),
-        }))
-        .then(certificate => {
-            // check that these key and certificate are valid, matching and can be loaded before using them
+// Get SSL certificates from local memroy.
+async function get_ssl_certificates() {
+    if (certs === null) {
+        try {
+            certs = await read_ssl_certificates();
+            using_generated_certs = false;
+            dbg.log0('Using mounted certificates');
+
+        } catch (err) {
+            if (err.code !== 'ENOENT') {
+                dbg.error('One or more SSL certificates failed to load', err.message);
+                dbg.warn('Fallback to generating self-signed certificates...');
+            }
+
+            dbg.warn('Generating self-signed certificates');
+            certs = Object.keys(SSL_CERTS_DIR_PATHS)
+                .reduce((c, service_name) => {
+                        c[service_name] = generate_ssl_certificate();
+                        return c;
+                }, {});
+            using_generated_certs = true;
+        }
+    }
+
+    return certs;
+}
+
+async function get_ssl_certificate(service) {
+    const { [service]: certificate = null } = await get_ssl_certificates();
+    if (certificate === null) {
+        throw new Error(`Invalid service name, got: ${service}`);
+    }
+
+    return certificate;
+}
+
+// Read SSL certificates form disk
+function read_ssl_certificates() {
+    return P.props(
+        _.mapValues(SSL_CERTS_DIR_PATHS, async dir => {
+            const key_path = path.join(dir, 'tls.key');
+            const cert_path = path.join(dir, 'tls.crt');
+            const certificate = {
+                key: await fs.promises.readFile(key_path, 'utf8'),
+                cert: await fs.promises.readFile(cert_path, 'utf8')
+            };
+
             verify_ssl_certificate(certificate);
```

Comment:
Not sure it's clear what is the behavior when the provided certificate file is invalid - do we fallback to generated cert? or do we crash the server?